### ISSUE TITLE
Blacklight::Base not invokable/usable

### ISSUE THREAD

**Issue by atz** [projectblacklight/blacklight#1155]
Tested against `5.9.3` but likely affecting other versions.  `Blacklight::Base` is a module, designed to be included in other code.  But...

## Ultrabasic
```ruby
module Bar
  include Blacklight::Base
end
```
gets:
```
NoMethodError: undefined method `class_attribute' for Bar:Module
from ~/.rvm/gems/ruby-1.9.3-p547/gems/blacklight-5.9.3/lib/blacklight/request_builders.rb:18:in `block in <module:RequestBuilders>'
```

## Variant
```ruby
class MyBase
  include Blacklight::Base
end  
```
gets:
```
NoMethodError: undefined method `helper_method' for MyBase:Class
from ~/.rvm/gems/ruby-1.9.3-p547/gems/blacklight-5.9.3/lib/blacklight/catalog/search_context.rb:8:in `block in <module:SearchContext>'
```

So what exactly is `Blacklight::Base` a Base for?  It appears it actually needs class_attribute stuff at the time it is included, but then it fails for other reasons when included from a class.

**Comment by jkeck** [comment]
@atz if I'm not mistaken it's meant to be included in a controller (or a module that is mixed into a controller).

https://github.com/sul-dlss/spotlight/blob/master/lib/spotlight/base.rb#L5
https://github.com/sul-dlss/spotlight/blob/master/app/controllers/spotlight/pages_controller.rb#L7

**Comment by jkeck** [comment]
And you can see where Blacklight mixes `Blacklight::Base` into the [`Catalog` module](https://github.com/projectblacklight/blacklight/blob/master/lib/blacklight/catalog.rb#L13) (which is included in the [`CatalogController`](https://github.com/projectblacklight/blacklight/blob/master/app/controllers/catalog_controller.rb#L4)

**Comment by cbeer** [comment]
Maybe we should require `active_support/core_ext/class/attribute` in `search_context` (assume that's the only expectation broken)? That'd at least make it usable elsewhere..

I do wonder what behavior you're trying to get out of `Blacklight::Base`, though, and if there's a better thing for you to include.

**Comment by jcoyne** [comment]
I think Joe's second usage example was actually feasible up until this commit: 630a77ab8a11bbde95eee2d4d65e7b81050fe67a

**Comment by atz** [comment]
@jkeck: by definition, if it requires a higher level class to be mixed into, it isn't a base anymore.  And should be sniffed out at include-time, or at least documented.

What I want is a single point of access in my app for the big cumbersome blacklight solr configuration, i.e. to wrap Blacklight::Base with my own app's base class so that everything can access it (read it).  And I want it to initialize at server spin-up and not per controller (or model!) per session.

If I try to follow the pattern of the other apps, I do:
```ruby
module Zyzz
  extend ActiveSupport::Concern
  include Blacklight::Base
end
class Foobar
  include Zyzz
end

## FAIL: NoMethodError: undefined method `helper_method' for Foobar:Class -lib/blacklight/catalog/search_context.rb:8
## But...

class Foobar
  include Zyzz
end
```
And then that succeeds!  THE EXACT SAME LINES.  And it repeats.  For each simple class that just includes the Base-wrapper, it fails to be defined the first time and succeeds the second.  (This is in rails console, obviously.)  That smells like weirdo side-effecting and obfuscated dependency-timing to me.

**Comment by jkeck** [comment]
That's definitely strange wrt the second class definition not producing the same error as the first.  Maybe somebody w/ more knowledge on how these classes get loaded can speak to why that might happen.

I'm not sure what the impetus for the `Blacklight::Base` module naming/pattern was, but it may have something to do w/ the nature of how Blacklight is expected to be extended as a rails engine (again, perhaps somebody w/ more knowledge in this pattern could speak to that).

As far as a ~~"solution"~~ workaround to do what you want (although I don't know if this is really any better) you could instantiate a `Blacklight::Configuration` object w/ your base configurations on application initialization (application.rb, an initializer, whatever).

```ruby
# application_blacklight_config_initializer.rb
SomethingGlobal.blacklight_config = Blacklight::Configuration.new do |config|
  config.default_solr_params = { 
    :qt => 'search',
    :rows => 10 
  }
  ...
end
```

You can then set `blacklight_config` to `SomethingGlobal.blacklight_config` in your preferred manner for any controller that needs that global configuration.  Again, I'm not sure this is really an ideal solution, but it is a *"solution"*.  If what you want is to set a configuration in an initializer that will be autoloaded into any blacklight controller, that would be more involved.

So all that being said, if you have strong opinions about how this can be done better, a new major version of Blacklight is impending and this kind of change would be something certainly worth evaluating for that release.  Please feel free to send over ideas for a better architecture to use, or even submit a Pull Request with some changes that you think gets us to a better place.

**Comment by cbeer** [comment]
I'd still suggest `Blacklight::Base` is not what you really want. I'm not sure why `SearchContext` needs to be mixed in there either. Given the timestamp of 630a77a, it's probably worth looking through github.com/sul-dlss/spotlight for how it uses it.

**Comment by atz** [comment]
It may very well be that another path is better for my app, but `Blacklight::Base` behavior/expectations should still be fixed.  In the scope of this issue *that* is what I want.  Certainly the same class definitions should not variously fail or succeed depending on execution order.  

The `Zyzz` example I posted was extrapolated directly from the way `Spotlight::Base` uses `Blacklight::Base`.  Specifically:
```ruby
module ...
 extend ActiveSupport::Concern
 include Blacklight::Base
```

Does Spotlight avoid error because it overrides blacklight_config() or because it always includes that base in *another* module layer that has necessary dependencies before including that in a class or because it has other precursory magic?  Otherwise, I don't see what instruction I am supposed to gain from Spotlight's example, since I don't want to override blacklight_config.

**Comment by cbeer** [comment]
In #1265, I've proposed moving `Blacklight::Base` into `app/controllers/concerns`, which should clarify its intended purpose.

**Comment by cbeer** [comment]
Fixed by #1265
